### PR TITLE
813 ah net

### PR DESCRIPTION
**Pull request by yiheng-wang-nv**
Implemented #813 .

### Description
AHNet and its necessary block FCN have been implemented. Based on the author's original implementations ( https://github.com/lsqshr/AH-Net ), one more deterministic version has been added. Specifically, in this version, nn.ConvTranspose3d is used to replace all interpolate and up-sampling manipulations. Moreover, the kernel size for the maxpool layer has been changed to meet the consistency with stride value which is necessary to avoid utilizing atomicAdd operation (an non-deterministic operation). 
Several control tests and ablation tests have been done to figure out the non-deterministic places, thus I think my conclusions are reliable. In addition, there are some evidences to support them: 
1. Pytorch's official explanation:
https://pytorch.org/docs/stable/notes/randomness.html#pytorch 

2. Similar findings in pooling layer's non-deterministic issue:
https://discuss.pytorch.org/t/inconsistent-results-with-3d-maxpool-on-gpu/38558/2

Since 2D model (FCN) uses resnet50 as its backbone and we do not wanna re-implement it, torchvision package is necessary. Whereas, the version of torchvision and pytorch has to be consistent. In MONAI, we use both Pytorch 1.4 (which needs torchvision 0.5) and the latest version (like Pytorch 1.6 which needs torchvision 0.7), thus my solution to maintain the compatibility is to modify several yml files in .github/workflows (please see the changed files for details). If there are some other more suitable strategies, feel free to mention here and I am able to change it.

What's more, this PR includes some changes for existing codes, including:

1. Some typing hints for SENet.
2. Some automatically modified formats by isort.
3. Re-write dropout layer related codes.
4. Added **inplace=True** in relu function in SENet.

All modifications can be traced in the corresponding commits with comments.

### Status
**Ready**

### Types of changes

- [ ] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [ ] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/networks/blocks/fcn.py [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Type
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torchvision import models
```
Comment (wyli) [inline monai/networks/blocks/fcn.py:17]: Hi @yiheng-wang-nv please make torchvision optional using `optional_import`. it'll be something like:
```python
from monai.utils import exact_version, optional_import
models, _ = optional_import("torchvision", "0.5.0", exact_version, "models")
```

**Comment by Nic-Ma** [comment]
Hi @wyli ,

As we don't use `torchvision` at other places, can we host the pre-trained model on dropbox to get rid of torchvision?
Thanks.

File: monai/networks/blocks/fcn.py (hunk 1 above)
Comment (yiheng-wang-nv) [reply monai/networks/blocks/fcn.py:17]: Sorry, I wrongly clicked the request re-review button, plz ignore that. Let me think about how to use the pretrained model in a easier way

**Comment by wyli** [comment]
> Hi @wyli ,
> 
> As we don't use `torchvision` at other places, can we host the pre-trained model on dropbox to get rid of torchvision?
> Thanks.

looks like the current code requires both a `.pth` file and a python script to define the network https://github.com/pytorch/vision/blob/master/torchvision/models/resnet.py
if that's the case then we should use torchvision for now, we don't want to reimplement torchvision's resnet.py

**Comment by Nic-Ma** [comment]
OK, sounds good.
@yiheng-wang-nv , please update to use `optional import` as @wyli suggested.

Thanks.

**Comment by Nic-Ma** [comment]
Hi @yiheng-wang-nv ,

Please update `torchvision` related logic in below files:
1. https://github.com/Project-MONAI/MONAI/blob/master/docs/source/installation.md
2. https://github.com/Project-MONAI/MONAI/blob/master/requirements-dev.txt
3. https://github.com/Project-MONAI/MONAI/blob/master/.github/workflows/pythonapp.yml
4. https://github.com/Project-MONAI/MONAI/blob/master/tests/min_tests.py

Thanks.

**Comment by Nic-Ma** [comment]
Thanks for the update.
Hi @wyli and @ericspod ,

Could you please help review it?
Thanks.

**Comment by yiheng-wang-nv** [comment]
> Thanks for the update.
> Hi @wyli and @ericspod ,
> 
> Could you please help review it?
> Thanks.

Thanks Nic-ma, wyli and ericspod. I've updated the description to roughly explain all the changed files. If there are any issues, feel free to contact me, thanks!

File: monai/networks/nets/ahnet.py [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+            drop_rate=0.0,
+        )
+        noutdense4 = num_init_features + densegrowth * ndenselayer
+
+        self.psp = PVP(noutdense4, deter_flag)
+        self.final = Final(4 + noutdense4, 1, deter_flag)
+
+        # Initialise parameters
+        for m in self.modules():
+            if isinstance(m, conv3d_type) or isinstance(m, conv3d_trans_type):
```
Comment (wyli) [inline monai/networks/nets/ahnet.py:370]: this line could be `if isinstance(m, (conv3d_type, conv3d_trans_type))` (it's the same for a few other isinstance usages)

File: monai/networks/blocks/fcn.py [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+class FCN(nn.Module):
+    """
+    2D FCN network with 3 input channels. The small decoder is built
+    with the GCN and Refine modules.
+    The code is adapted from lsqshr's original version:
+    https://github.com/lsqshr/AH-Net/blob/master/net2d.py
+
+    Args:
+        nout (int): number of output channels. Defaults to 1.
+        deter_flag (bool): let deter_flag=True if needs to make the training process deterministic.
```
Comment (wyli) [inline monai/networks/blocks/fcn.py:109]: `deter_flag` will only affect the upsampling in this particular module, how about changing `deter_flag` to `upsample_mode` and making comments that some choices of `upsample_mode` couldn't provide reproducibility? it becomes more intuitive and in the future we may extend the upsampling mode (it's the same logic for a few other modules that have `deter_flag`

File: monai/networks/nets/ahnet.py [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+        self.add_module(
+            "conv",
+            conv3d_type(
+                num_input_features, num_output_features, kernel_size=(3, 3, 1), stride=1, padding=(1, 1, 0), bias=False
+            ),
+        )
+        if deter_flag:
+            conv3d_trans_type: Type[nn.ConvTranspose3d] = Conv[Conv.CONVTRANS, 3]
+            self.add_module(
+                "pool", conv3d_trans_type(num_output_features, num_output_features, kernel_size=2, stride=2, bias=False)
```
Comment (wyli) [inline monai/networks/nets/ahnet.py:146]: upsampling is named 'pool'? is this a typo?

File: monai/networks/nets/ahnet.py [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+        self.dense4 = DenseBlock(
+            num_layers=ndenselayer,
+            num_input_features=num_init_features,
+            bn_size=densebn,
+            growth_rate=densegrowth,
+            drop_rate=0.0,
+        )
+        noutdense4 = num_init_features + densegrowth * ndenselayer
+
+        self.psp = PVP(noutdense4, deter_flag)
```
Comment (wyli) [inline monai/networks/nets/ahnet.py:365]: looks like this is a PSP module https://arxiv.org/pdf/1612.01105.pdf, not sure why it is called PVP, need a reference for the PVP module if possible

File: monai/networks/blocks/fcn.py [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+        self.refine10 = Refine(self.nout)
+        self.transformer = self.conv2d_type(in_channels=256, out_channels=64, kernel_size=1)
+
+        if self.deter_flag:
+            conv2d_trans_type: Type[nn.ConvTranspose2d] = Conv[Conv.CONVTRANS, 2]
+            self.up_conv = conv2d_trans_type(
+                in_channels=self.nout, out_channels=self.nout, kernel_size=2, stride=2, bias=False,
+            )
+
+    def _regresser(self, inplanes: int) -> nn.Sequential:
```
[the remark this answers is not in the document]
Comment (yiheng-wang-nv) [reply monai/networks/blocks/fcn.py:162]: Yes, I've removed it.

File: monai/networks/blocks/fcn.py [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+    """
+
+    def __init__(self, nin: int = 3, nout: int = 1, deter_flag: bool = False):
+        super(MCFCN, self).__init__(nout=nout, deter_flag=deter_flag)
+
+        relu_type: Type[nn.ReLU] = Act[Act.RELU]
+        conv2d_type: Type[nn.Conv2d] = Conv[Conv.CONV, 2]
+        norm2d_type: Type[nn.BatchNorm2d] = Norm[Norm.BATCH, 2]
+
+        self.init_proj = nn.Sequential(
```
[the remark this answers is not in the document]
Comment (yiheng-wang-nv) [reply monai/networks/blocks/fcn.py:234]: Thanks, I modify it.

File: monai/networks/nets/ahnet.py (hunk 2 above)
Comment (yiheng-wang-nv) [reply monai/networks/nets/ahnet.py:370]: I see, I've changed them all.

File: monai/networks/nets/ahnet.py (hunk 4 above)
Comment (yiheng-wang-nv) [reply monai/networks/nets/ahnet.py:146]: It comes from the author's official code: see https://github.com/lsqshr/AH-Net/blob/master/net3d.py  (line 195), can I change it into "up"?

File: monai/networks/nets/ahnet.py (hunk 5 above)
Comment (yiheng-wang-nv) [reply monai/networks/nets/ahnet.py:365]: It comes from the author's official code: see https://github.com/lsqshr/AH-Net/blob/master/net3d.py (line 267), use PSP makes more sense I think, let me modify it.

File: monai/networks/blocks/fcn.py (hunk 3 above)
Comment (yiheng-wang-nv) [reply monai/networks/blocks/fcn.py:109]: I see, I've updated.

File: .github/workflows/pythonapp.yml [hunk 8]
```diff
@@ -162,7 +164,7 @@ jobs:
         # install the latest pytorch for testing
         # however, "pip install monai*.tar.gz" will build cpp/cuda with an isolated
         # fresh torch installation according to pyproject.toml
-        python -m pip install torch>=1.4
+        python -m pip install torch>=1.4 torchvision
```
[the remark this answers is not in the document]
Comment (Nic-Ma) [reply .github/workflows/pythonapp.yml:167]: Hi @wyli ,

May I know why here we use "torch>=1.4" here?
Can we use `python -m pip install torch==1.4 torchvision==0.5.0` same as other places?

Thanks.

File: monai/networks/nets/ahnet.py (hunk 4 above)
Comment (wyli) [reply monai/networks/nets/ahnet.py:146]: sure thanks

File: .github/workflows/pythonapp.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (wyli) [reply .github/workflows/pythonapp.yml:167]: that is to be consistent with https://github.com/Project-MONAI/MONAI/blob/663945f07a546bb8d897e94e33222654fb7f1698/requirements.txt#L1

File: monai/networks/blocks/convolutions.py [hunk 9]
```diff
@@ -98,7 +103,7 @@ def __init__(
             else:
                 drop_name, drop_args = split_args(dropout)
 
-            drop_type = Dropout[drop_name, dimensions]
+            drop_type = Dropout[drop_name, dropout_dim]
```
[the remark this answers is not in the document]
Comment (yiheng-wang-nv) [reply monai/networks/blocks/convolutions.py:106]: Thanks, I've added a validation.

File: .github/workflows/pythonapp.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (yiheng-wang-nv) [reply .github/workflows/pythonapp.yml:167]: Hi @wyli , here could I remove the "exact_version" parameter? The version of `torchvision` should be consistent with the version of `torch`, thus here I added the installing commands after `pip install torch` in order to face our different versions' check.
`torchvision` has been added to `requirements-dev.txt`. Could you please take a look at the updates? Thanks : )

[the remark this answers is not in the document]
Comment (wyli) [reply .github/workflows/pythonapp.yml:167]: sure, sounds good to me
